### PR TITLE
feat: add devcontainer

### LINKED ISSUE #23: Proof-of-Concept for Containerized Environment for Kit
I couldn't fork or make a PR for this so here is an issue that presents a working **prototype**.

It is based on this image:
* https://github.com/accetto/ubuntu-vnc-xfce-g3/blob/master/docker/xfce-firefox/README.md 

## Dockerfile

```
FROM accetto/ubuntu-vnc-xfce-firefox-g3

USER root

RUN sudo apt update
RUN sudo apt-get -y install wget gpg
RUN sudo wget -qO- https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > packages.microsoft.gpg
RUN sudo install -o root -g root -m 644 packages.microsoft.gpg /etc/apt/trusted.gpg.d/
RUN sudo sh -c 'echo "deb [arch=amd64,arm64,armhf signed-by=/etc/apt/trusted.gpg.d/packages.microsoft.gpg] https://packages.microsoft.com/repos/code stable main" > /etc/apt/sources.list.d/vscode.list'
RUN sudo rm -f packages.microsoft.gpg
RUN sudo apt -y install apt-transport-https
RUN sudo apt update
RUN sudo apt -y install code

USER headless
```

## To create the image:
`docker build -t tiger .`

## To run the image:
`docker run -p 6901:6901 tiger`

## Connect to noVNC:
* Point browser at: `localhost:6901`
* Click "Connect"
* Password: headless

## Launch VSCode:
* Open a terminal
* `code --no-sandbox`

## Comments
* This needs some polish but it works pretty well and is very responsive on my machine.
  * Make a nice user
  * Install git 
  * etc...
* I started from an image that has TigerVNC installed
  * It runs both a VNC and a noVNC server.
  * It is being actively maintained with a new image within the last two weeks. 
* One challenge is copy and paste between host and container
  * There is a clipboard hack in the noVNC client that works but is inconvenient.
  * The TigerVNC client works better, allows host/container copy/paste and is simple to install.
    *  https://tigervnc.org/
    * Need to open another port to use the standalone client.
      * `docker run -p 6901:6901 -p 5901:5901 tiger`
    * The UI appears to be in french when I run the most recent version but I suspect this can be fixed.
* The need to use `--no-sandbox` flag on VSCode is a quick hack to fix the issue:
  * `Failed to move to new namespace: PID namespaces supported, Network namespace supported, but failed: errno = Operation not permitted`
  * Should be pretty easily fixable with some research.

### PR DESCRIPTION
**Pull request by StoneyJackson**
This is the start of a fully contained development environment.

The goal is to build and push a Docker image that contains a
complete development environment. The only dependency would be
Docker. Then with a command like...

```
$ docker run url-to-image
...
...
...
Open a browser to http://localhost:35000/
```

Then the user does all their work inside that browser.

Co-authored-by: Grant Braught <braught@dickinson.edu>

---

When merged, closes #23 and #24 

---
__Licensing Certification__

FarmData2 is a [Free Cultural Work](https://freedomdefined.org/Definition) and all accepted contributions are licensed as described in the LICENSE.md file. This requires that the contributor holds the rights to do so. By submitting this pull request __I certify that I satisfy the terms of the [Developer Certificate of Origin](https://developercertificate.org/)__ for its contents.

### CODE REVIEW

**Comment by StoneyJackson** [comment]
Start with /docker/gitkit-devcontainer/README.md

**Comment by StoneyJackson** [comment]
I want to squash merge this thing to get rid of all the wip: commits. So please do not merge, just review. And if and when we are satisfied, I'll "merge" it.

**Comment by StoneyJackson** [comment]
@cmacdonell Ready for testing. Specifically

* [HFOSS/GitKitTest](https://github.com/HFOSSedu/GitKitTest) now exists.
* The latest image of [hfossedu/gitkit-devenv](https://hub.docker.com/repository/docker/hfossedu/gitkit-devenv) has been pushed to Docker Hub.
* The latest instructions for using hfossedu/gitkit-devenv are in this branch (24-build-container-for-gitkit) in ./docker/gitkit-devenv/README.md .

Please let me know if you find any problems or can think of any improvements.

**Comment by cmacdonell** [comment]
When I try to run "code" in the devenv, it is not found.

File: docker/gitkit-devenv/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        meld
+
+# Install VS Code
+RUN apt-get -y install wget gpg apt-transport-https && \
+    wget -qO- https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > packages.microsoft.gpg && \
+    install -o root -g root -m 644 packages.microsoft.gpg /etc/apt/trusted.gpg.d/ && \
+    sh -c 'echo "deb [arch=amd64,arm64,armhf signed-by=/etc/apt/trusted.gpg.d/packages.microsoft.gpg] https://packages.microsoft.com/repos/code stable main" > /etc/apt/sources.list.d/vscode.list' && \
+    rm -f packages.microsoft.gpg && \
+    apt-get update && \
+    apt-get -y install code && \
```
Comment (cmacdonell) [inline docker/gitkit-devenv/Dockerfile:22]: I see this is here, but vscode appears not be getting installed.

Comment (StoneyJackson) [reply docker/gitkit-devenv/Dockerfile:22]: OK, it looks like code depends on gpg, so the remove at the end was causing code to be removed as well.

Not sure how I missed this. But I think I can fix it. Give me a bit.

Comment (StoneyJackson) [reply docker/gitkit-devenv/Dockerfile:22]: The new image should be up. Please try it again.

**Comment by StoneyJackson** [comment]
> FWIW, `docker exec -it devenv /bin/bash` works to shell into the devenv from the host computer. Docker Desktop's CLI is some unknown shell missing features like tab-completion. This command gives bash.

@braughtg  @cmacdonell

Is this a show stopper for POSSE? Or can we live with this?

**Comment by cmacdonell** [comment]
No, certainly not a showstopper. I offer it just in case it's helpful. The noVNC browser interface works just fine. The only downside is copying-and-pasting the github repo name is a nuisance via noVNC, so they'll have a type it in or copy it from a browser within the devenv.

File: docker/gitkit-devenv/Dockerfile (hunk 1 above)
Comment (cmacdonell) [reply docker/gitkit-devenv/Dockerfile:22]: Using the latest image, vscode is now there.

**Comment by StoneyJackson** [comment]
OK. I think we're good. I'm going to merge.

**Pull request merged**